src/borg/archiver.py

```diff
@@ -3125,7 +3125,10 @@ def run(self, args):
         self.prerun_checks(logger)
         if is_slow_msgpack():
             logger.warning("Using a pure-python msgpack! This will result in lower performance.")
-        return func(args)
+        exit_code = func(args)
+        if is_warning_printed():
+            return EXIT_WARNING
+        return exit_code
```

Comment:
if exit_code is already EXIT_ERROR, but a  warning was printed, then it would become EXIT_WARNING.